Framework/API/src/Algorithm.cpp

```diff
@@ -1651,6 +1656,31 @@ bool Algorithm::getAlgStartupLogging() const {
   return m_isAlgStartupLoggingEnabled;
 }
 
+void Algorithm::declareProperty(std::unique_ptr<Kernel::Property> p,
+                                const std::string &doc) {
+  if (isCompoundProperty(p->name()))
+    throw std::runtime_error(
+        p->name() +
+        " has already been used by Algorithm::declareIndexProperty.");
+
+  PropertyManagerOwner::declareProperty(std::move(p), doc);
+}
+
+Kernel::IPropertyManager::TypedValue
+Algorithm::getProperty(const std::string &name) const {
+  if (isCompoundProperty(name))
+    throw std::runtime_error("Algorithm::getIndexProperty must be used with "
+                             "properties declared using "
+                             "Algorithm::declareIndexProperty.");
+
+  return PropertyManagerOwner::getProperty(name);
+}
+
+bool Algorithm::isCompoundProperty(const std::string &name) const {
+  return std::find(m_reservedList.cbegin(), m_reservedList.cend(), name) !=
```

Comment:
`declareProperty` currently throws if you try and declare something that already has that name so we are safe there.